Skip UTF-8 BOM bytes from Dockerfile and .dockerignore if exist

yongtang opened the pull request:
**\- What I did**

This fix tries to address issues in #23221 where Dockerfile or .dockerignore may consists of UTF-8 BOM. This likely happens when Notepad tries to save a file as UTF-8 in Windows.

**\- How I did it**

This fix skips the UTF-8 BOM bytes from the beginning of the Dockerfile or .dockerignore if exist.

**\- How to verify it**

Additional tests has been added to cover the changes in this fix.

**\- Description for the changelog**

Skip UTF-8 BOM bytes from the beginning of the Dockerfile or .dockerignore if exist.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #23221.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

yongtang committed:
Skip UTF-8 BOM bytes from Dockerfile if exists

This fix tries to address issues in #23221 where Dockerfile
may consists of UTF-8 BOM. This likely happens when Notepad
tries to save a file as UTF-8 in Windows.

This fix skips the UTF-8 BOM bytes from the beginning of the
Dockerfile if exists.

Additional tests has been added to cover the changes in this
fix.

This fix fixes #23221.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

justincormack commented:
cc @jhowardmsft

justincormack commented:
I think this is a good idea, as I have seen the pain of random things failing for this reason in the past and being asked to debug. It should I guess also apply to `.dockerignore`?

thaJeztah commented on builder/dockerfile/parser/parser.go:
```diff
@@ -152,8 +153,14 @@ func Parse(rwc io.Reader) (*Node, error) {
-		scannedLine := strings.TrimLeftFunc(scanner.Text(), unicode.IsSpace)
+		scannedBytes := scanner.Bytes()
+		// We trim UTF8 BOM
+		if currentLine == 0 && bytes.HasPrefix(scannedBytes, utf8bom) {
```
the `bytes.HasPrefix()` is redundant, because `TrimPrefix()` already does this; https://golang.org/src/strings/strings.go?s=16765:16805#L612

yongtang replied:
Thanks @thaJeztah The pull request has been updated to remove the `bytes.HasPrefix()`.

thaJeztah commented:
Got bit by this a few times, sgtm

allencloud commented:
Great. Thanks @yongtang

yongtang committed:
Skip UTF-8 BOM bytes from Dockerignore if exist

This fix tries to address issues related to #23221 where Dockerignore
may consists of UTF-8 BOM. This likely happens when Notepad
tries to save a file as UTF-8 in Windows.

This fix skips the UTF-8 BOM bytes from the beginning of the
Dockerignore if exists.

Additional tests has been added to cover the changes in this fix.

This fix is related to #23221 (UTF-8 BOM in Dockerfile).

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @justincormack @thaJeztah @allencloud for the feedback. I updated the pull request. Now UTF-8 BOM in bother Dockerfile and .dockerignore will be skipped. Let me know if there are any issues.

jhowardmsft commented:
LGTM

thaJeztah commented:
LGTM failing test is https://github.com/docker/docker/issues/23246

thaJeztah commented:
Windows failure is not related, and will be fixed in https://github.com/docker/docker/issues/23246, so ignoring that failing test
